src/libopensc/pkcs15-cardos.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * PKCS15 emulation layer for CardOS cards
+ * Adapted from PKCS15 emulation layer for IAS/ECC card.
+ *
+ * Copyright (C) 2020, Douglas E. Engert <DEEngert@gmail.com>
+ * Copyright (C) 2016, Viktor Tarasov <viktor.tarasov@gmail.com>
+ * Copyright (C) 2004, Bud P. Bruegger <bud@comune.grosseto.it>
+ * Copyright (C) 2004, Antonino Iacono <ant_iacono@tin.it>
+ * Copyright (C) 2003, Olaf Kirch <okir@suse.de>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "internal.h"
+#include "pkcs15.h"
+
+
+/*
+ * Called after sc_pkcs15_bind_internal
+ * Create new flags based on supported_algos.
+ */
+static int cardos_fix_token_info(sc_pkcs15_card_t *p15card)
+{
+	sc_card_t *card;
+	struct sc_supported_algo_info (*saa)[SC_MAX_SUPPORTED_ALGORITHMS];
+	struct sc_supported_algo_info *sa;
+	struct sc_cardctl_cardos_pass_algo_flags *passed = NULL;
+	int r = 0;
+	int i;
+
+	card = p15card->card;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	passed = calloc(1, sizeof(struct sc_cardctl_cardos_pass_algo_flags));
+	if (!passed)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_ENOUGH_MEMORY);
+
+	passed->pass = 1; /* get used_flags and card_flags from card */
+	r = sc_card_ctl(p15card->card, SC_CARDCTL_CARDOS_PASS_ALGO_FLAGS, passed);
+	if (r < 0) {
+		free(passed);
+		LOG_FUNC_RETURN(card->ctx, r);
+	}
+
+	saa = &(p15card->tokeninfo->supported_algos);
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "Original Flags: 0x%8.8lx card->flags:0x%8.8lx", passed->used_flags, passed->card_flags);
+
+	if (passed->card_flags) { /* user forced the flags, use them */
+		passed->new_flags = passed->card_flags; /* from card_atr flags */
+	} else {
+
+		for (i = 0, sa = saa[0]; i < SC_MAX_SUPPORTED_ALGORITHMS; i++, sa++) {
+
+			if (sa->reference == 0 && sa->reference == 0 && sa->mechanism == 0
+					&& sa->operations == 0 && sa->algo_ref == 0)
+				break;
+
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "supported_algos[%d] mechamism:0x%8.8x", i, sa->mechanism);
+			switch(sa->mechanism) {
+			case 0x01 :
```

Comment:
clang-format suggested the following change:


```
-			switch(sa->mechanism) {
-			case 0x01 :
+			switch (sa->mechanism) {
+			case 0x01:
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/pkcs15-cardos.c -->